### PR TITLE
feat(iorails): Tool-calling - streaming tools from main LLM

### PR DESCRIPTION
**Pull request by tgasser-nv**
## Description

This PR adds streaming tool-calling functionality to Guardrails as part of the PR stack below. No introspection of the tool calls or return values is done, just passing back and forth between the Main LLM to decide which tool calls to make, and how to respond based on prior conversation turns and tool results the harness executed.

* **Merged** #2016 : Non-streaming tool-calling connected to main LLM for inference and back to the client with the response. No canonicalization.
* **This PR** : Streaming tool-calling implementation. Just propagating tool calls to the Main LLM and back again without canonicalization..
* #2030  : Add canonical internal dataclasses for tool-calling. Implement tool-calling rails to check the LLM-generated function call matches the provided tool signature and schema , and that the response from the function execution matches the signature and schema.
* **TODO** : Docs PR describing the feature as implemented.

## Related Issue(s)

Stacked PR on top of #2016 (which added non-streaming tool-calling).

### Pre-commit

```
$ poetry run pre-commit run --all-files
check yaml...............................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
ruff (legacy alias)......................................................Passed
ruff format..............................................................Passed
Insert license in comments...............................................Passed
pyright..................................................................Passed
```

### Unit-test

```
$ make test
...............ss....ss......s.........ss....s................................................................. [  2%]
............................................................................................................... [  4%]
......s..........................................................sss...s....................................... [  6%]
............................................................................................................... [  9%]
........................................................................s...................................... [ 11%]
............................................................................................................... [ 13%]
............................................................................................................... [ 16%]
............................................................................................................... [ 18%]
............................................................................................................... [ 20%]
.............................................................................s.........s.s..................... [ 22%]
.............ss.sssss.....................................................s.................................... [ 25%]
..............................s................................................................................ [ 27%]
............................................................................................................... [ 29%]
....s.ssssss.s............sss.sss...ss..s................s.sssssss............................................. [ 32%]
............................................................................................................... [ 34%]
............................................................................................................... [ 36%]
....................................ss......................................................................... [ 38%]
............................................................................................................... [ 41%]
.................................sssss...........................s.....................................sssssss. [ 43%]
.......................sssss.ssss..sssssss.ss....................ss............................................ [ 45%]
...........s............s.s...............................ss..................................s................ [ 48%]
............s.ssss............................................................................................. [ 50%]
.................s.ssss........................................................................................ [ 52%]
.................................................................s............................................. [ 54%]
............................................................................................................... [ 57%]
.................s............................................................................................. [ 59%]
............................................................................................................... [ 61%]
............................................................................................................... [ 64%]
............s.............................................................s.................................... [ 66%]
............................................................................................................... [ 68%]
.............................................................................................................ss [ 71%]
............................................................................................................... [ 73%]
...............................................................................s..........................sssss [ 75%]
ss.sss.................s.s..................................................................................... [ 77%]
.......................................................ssssssss................................................ [ 80%]
............................................................................................................... [ 82%]
..................sssssss...........................sss........................................................ [ 84%]
....................................s......s..ss......ss....................................................... [ 87%]
..............................................................................................s................ [ 89%]
........................................................sssssss..ss.ssssssssss................................. [ 91%]
......................................sss...................................s.................................. [ 93%]
...............................sss.sss.ss...............................................ss..................... [ 96%]
.......................................................................................................s....... [ 98%]
........................................................................                                        [100%]
========================================= 4665 passed, 180 skipped in 41.69s ==========================================
```

### Integration test (Using [`e2e_streaming_main_llm.py`](https://gist.github.com/tgasser-nv/e1bf5a10b28f75a85aba051f3554172e))

See attached log file for all INFO and DEBUG logging, only INFO lines are listed below. This integration script uses a dummy tool to find the weather conditions in Paris. It checks on both a NIM-model `meta/llama-3.3-70b-instruct` and an OpenAI model `gpt-4o-mini`. 

Here's the full log, with all chunks logged out:

[20260613_streaming_tool_calls_raw_chunks.txt](https://github.com/user-attachments/files/28929578/20260613_streaming_tool_calls_raw_chunks.txt)

```
$ poetry run python ~/utils/toolcalling/e2e_streaming_main_llm.py --debug  |& tee ~/logs/20260613_streaming_tool_calls_raw_chunks.txt

2026-06-14 10:01:35 INFO: Registered model engine: type=main, model=gpt-4o-mini, base_url=https://api.openai.com
2026-06-14 10:01:35 INFO: RailsManager initialized: input_flows=[], output_flows=[], input_parallel=False, output_parallel=False
2026-06-14 10:01:35 INFO: [5235496e0538eace] stream_async called
2026-06-14 10:01:35 INFO: [5235496e0538eace] Running input rails
2026-06-14 10:01:35 INFO: [5235496e0538eace] Streaming main LLM
2026-06-14 10:01:35 INFO: [5235496e0538eace] HTTP POST (stream) https://api.openai.com/v1/chat/completions model='gpt-4o-mini'
2026-06-14 10:01:42 INFO: [5235496e0538eace] Tool-call-only stream: output rails skipped
2026-06-14 10:01:42 INFO: [5235496e0538eace] generation task completed time=6659.2ms
2026-06-14 10:01:42 INFO: [5235496e0538eace] stream_async completed time=6660.0ms
2026-06-14 10:01:42 INFO: [57f026a937dc19fe] stream_async called
2026-06-14 10:01:42 INFO: [57f026a937dc19fe] Running input rails
2026-06-14 10:01:42 INFO: [57f026a937dc19fe] Streaming main LLM
2026-06-14 10:01:42 INFO: [57f026a937dc19fe] HTTP POST (stream) https://api.openai.com/v1/chat/completions model='gpt-4o-mini'
2026-06-14 10:01:48 INFO: [57f026a937dc19fe] generation task completed time=6459.1ms
2026-06-14 10:01:48 INFO: [57f026a937dc19fe] stream_async completed time=6459.4ms
2026-06-14 10:01:48 INFO: Registered model engine: type=main, model=meta/llama-3.3-70b-instruct, base_url=https://integrate.api.nvidia.com
2026-06-14 10:01:48 INFO: RailsManager initialized: input_flows=[], output_flows=[], input_parallel=False, output_parallel=False
2026-06-14 10:01:48 INFO: [5dad0055b1a0f908] stream_async called
2026-06-14 10:01:48 INFO: [5dad0055b1a0f908] Running input rails
2026-06-14 10:01:48 INFO: [5dad0055b1a0f908] Streaming main LLM
2026-06-14 10:01:48 INFO: [5dad0055b1a0f908] HTTP POST (stream) https://integrate.api.nvidia.com/v1/chat/completions model=''
2026-06-14 10:01:55 INFO: [5dad0055b1a0f908] Tool-call-only stream: output rails skipped
2026-06-14 10:01:55 INFO: [5dad0055b1a0f908] generation task completed time=7377.5ms
2026-06-14 10:01:55 INFO: [5dad0055b1a0f908] stream_async completed time=7378.1ms
2026-06-14 10:01:55 INFO: [a9da9ed549aebbf8] stream_async called
2026-06-14 10:01:55 INFO: [a9da9ed549aebbf8] Running input rails
2026-06-14 10:01:55 INFO: [a9da9ed549aebbf8] Streaming main LLM
2026-06-14 10:01:55 INFO: [a9da9ed549aebbf8] HTTP POST (stream) https://integrate.api.nvidia.com/v1/chat/completions model='meta/llama-3.3-70b-instruct'
2026-06-14 10:02:12 INFO: [a9da9ed549aebbf8] generation task completed time=17050.4ms
2026-06-14 10:02:12 INFO: [a9da9ed549aebbf8] stream_async completed time=17050.8ms
INFO: text streamed before the tool call: ''
INFO: answer mentions returned temperature ('18'): True
INFO: text streamed before the tool call: ''
INFO: answer mentions returned temperature ('18'): True

```

## Checklist

- [ ] I've read the [CONTRIBUTING](https://github.com/NVIDIA-NeMo/Guardrails/blob/develop/CONTRIBUTING.md) guidelines.
- [ ] This PR links to a triaged issue assigned to me.
- [ ] My PR title follows the project commit convention.
- [ ] I've updated the documentation if applicable.
- [ ] I've added tests if applicable.
- [ ] I've noted any verification beyond CI and any checks I couldn't run.
- [ ] I did not update generated changelog files manually.
- [ ] I addressed all CodeRabbit, Greptile, and other review comments, or replied with why no change is needed.
- [ ] @mentions of the person or team responsible for reviewing proposed changes.

## Summary by CodeRabbit

* **New Features**
  * Tool calls can now be streamed and emitted as a final JSON chunk, alongside text content when present.

* **Bug Fixes**
  * Improved handling of tool-call-only responses; output-rail safety checks are properly skipped when appropriate.

* **Tests**
  * Added comprehensive test coverage for tool-call streaming scenarios, including fragmented arguments assembly and edge cases.

### CODE REVIEW

File: nemoguardrails/guardrails/iorails.py [hunk 1]
```diff
@@ -756,6 +767,11 @@ async def _wrapped_iterator():
                                                     if isinstance(text, str) and text:
                                                         delivered.append(text)
                                             yield chunk
+                                    # Yield assembled tool calls as a terminal JSON chunk
+                                    # after all text content (and output rails) have
+                                    # finished.
+                                    if accumulated_tool_calls:
+                                        yield json.dumps({"tool_calls": _serialize_tool_calls(accumulated_tool_calls)})
```
[the remark this answers is not in the document]
Comment (tgasser-nv) [reply nemoguardrails/guardrails/iorails.py]: @greptile-apps Check latest code and make sure this has been fixed. Can this be resolved?

[the remark this answers is not in the document]
Comment (tgasser-nv) [reply nemoguardrails/guardrails/iorails.py]: @greptile-apps Check latest code and make sure this has been fixed. Can this be resolved?

[the remark this answers is not in the document]
Comment (tgasser-nv) [reply nemoguardrails/guardrails/iorails.py]: @coderabbitai Check latest code and make sure this has been fixed. Can this be resolved?

File: nemoguardrails/guardrails/model_engine.py [hunk 2]
```diff
@@ -212,6 +212,65 @@ def _parse_chat_completion_chunk(chunk: dict) -> Optional[LLMResponseChunk]:
...
+            }
+        else:
+            if tool_call_delta.get("id"):
+                tool_calls[index]["id"] = tool_call_delta["id"]
+            name = (tool_call_delta.get("function") or {}).get("name")
+            if name:
+                tool_calls[index]["name"] = name
+        arguments_fragment = (tool_call_delta.get("function") or {}).get("arguments") or ""
+        if arguments_fragment:
+            tool_calls[index]["arguments_buffer"] += arguments_fragment
```
[the remark this answers is not in the document]
Comment (tgasser-nv) [reply nemoguardrails/guardrails/model_engine.py:266]: @coderabbitai Check latest code and make sure this has been fixed. Can this be resolved?

File: nemoguardrails/guardrails/model_engine.py [hunk 3]
```diff
@@ -461,6 +523,10 @@ async def stream_call(
                     line = raw_line.decode("utf-8").strip()
                     if not line:
                         continue
+
+                    # Debug-log raw SSE line before any processing
+                    log.debug("[%s] SSE line: %s", req_id, line)
```
[the remark this answers is not in the document]
Comment (tgasser-nv) [reply nemoguardrails/guardrails/model_engine.py]: @coderabbitai Check latest code and make sure this has been fixed. Can this be resolved?

File: nemoguardrails/guardrails/iorails.py [hunk 4]
```diff
@@ -602,6 +621,9 @@ def stream_async(
             llm_kwargs = options.llm_params
 
         streaming_handler = StreamingHandler(include_metadata=include_metadata)
+        # Assembled tool calls from the stream — populated by _generation_task
+        # and consumed by _wrapped_iterator after the content stream ends.
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/iorails.py]: nit: is it documenting a data-flow coupling that only exists because of the shared-list design? can we add a FIXME or TODO?

File: nemoguardrails/guardrails/iorails.py [hunk 5]
```diff
@@ -629,14 +651,19 @@ async def _generation_task(request_span):
...
+                # stream ends. Reasoning is dropped for LLMRails compatibility.
                 log.info("[%s] Streaming main LLM", req_id)
                 content_parts: list[str] = []
                 async for chunk in self.engine_registry.stream_model_call("main", messages, **llm_kwargs):
                     if chunk.delta_content:
                         content_parts.append(chunk.delta_content)
                         await streaming_handler.push_chunk(chunk.delta_content)
+                    if chunk.delta_tool_calls:
+                        # Slice-assign to mutate the shared list in place so
+                        # _wrapped_iterator sees the update without nonlocal.
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/iorails.py]: nit: please make it TODO or FIXME, this is a cross-closure communication via mutable shared state which the comment is documenting it but better to document it as a potential smell that needs attention. If we refactor the channel (a small holder object, or nonlocal with an explicit assignment) then the comment becomes superfluous and we can drop it.

File: nemoguardrails/guardrails/model_engine.py [hunk 6]
```diff
@@ -461,6 +545,11 @@ async def stream_call(
                     line = raw_line.decode("utf-8").strip()
                     if not line:
                         continue
+
+                    # Debug-log raw SSE line before any processing. Bounded via
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/model_engine.py]: nit: please delete this

File: nemoguardrails/guardrails/model_engine.py [hunk 7]
```diff
@@ -461,6 +545,11 @@ async def stream_call(
...
+                    # truncate() for now to match the other payload logs.
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/model_engine.py]: nit: "for now"? let's change it to TODO or FIXME so that we can track

File: nemoguardrails/guardrails/model_engine.py [hunk 8]
```diff
@@ -212,6 +212,87 @@ def _parse_chat_completion_chunk(chunk: dict) -> Optional[LLMResponseChunk]:
...
+                    log.warning(
+                        "[%s] tool-call id %r collided with accumulator slot %d (existing id %r); "
+                        "provider likely omitted 'index' for parallel tool calls — result may be corrupted",
+                        get_request_id(),
+                        incoming_id,
+                        index,
+                        existing_id,
+                    )
+                tool_calls[index]["id"] = incoming_id
+            name = (tool_call_delta.get("function") or {}).get("name")
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/model_engine.py]: it seems that function delta rebuilt 3x in this func; but read once into a local. I haven't checked it though

File: nemoguardrails/guardrails/model_engine.py [hunk 9]
```diff
@@ -212,6 +212,87 @@ def _parse_chat_completion_chunk(chunk: dict) -> Optional[LLMResponseChunk]:
...
+            if name:
+                tool_calls[index]["name"] = name
+        arguments_fragment = (tool_call_delta.get("function") or {}).get("arguments") or ""
+        if arguments_fragment:
+            tool_calls[index]["arguments_buffer"] += arguments_fragment
+
+
+def _finalize_tool_calls(tool_calls: dict[int, dict]) -> list[ToolCall]:
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/model_engine.py:271]: duplicates openai_chat._finalize_tool_calls, should be OK for now better to add a TODO or FIXME so that we can get back to it later.

File: nemoguardrails/guardrails/model_engine.py [hunk 10]
```diff
@@ -474,10 +563,38 @@ async def stream_call(
...
+                    # A streamed tool call ends with a finish_reason — "tool_calls"
+                    # when the model chose freely, but "stop" when tool_choice forced
+                    # a specific function. Gate on *any* finish_reason (with tool calls
+                    # accumulated), not the literal "tool_calls", or forced calls never
+                    # surface. finish_reason keeps that chunk alive in
+                    # _parse_chat_completion_chunk, so parsed_chunk is non-None here; if a
+                    # provider ever omits a parseable finish frame, the post-loop safety
+                    # net below surfaces the calls instead.
+                    choices = raw_chunk.get("choices") or []
+                    finish_reason = choices[0].get("finish_reason") if choices else None
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/model_engine.py:591]: fires on any `finish_reason`; e.g. finish_reason="length" surfaces a truncated tool call with arguments={} silently (no warning).

File: nemoguardrails/guardrails/iorails.py [hunk 11]
```diff
@@ -629,14 +651,19 @@ async def _generation_task(request_span):
...
+                        accumulated_tool_calls[:] = chunk.delta_tool_calls
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/iorails.py]: in this case last-write-wins, not accumulation. we are losing earlier calls if engine emits delta_tool_calls on >1 chunk.

File: nemoguardrails/guardrails/iorails.py [hunk 12]
```diff
@@ -756,6 +792,24 @@ async def _wrapped_iterator():
                                                     if isinstance(text, str) and text:
                                                         delivered.append(text)
                                             yield chunk
+                                    # Yield assembled tool calls as a terminal chunk after
+                                    # all text content (and output rails) have finished —
+                                    # but only on a clean stream. If an error or guardrails
+                                    # violation was already surfaced, suppress the tool calls
+                                    # so the caller never receives a tool-call after a
+                                    # failure/block.
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/iorails.py]: nit: what-this-block-does comments. it is needed? we can make it explicit by extracting function(s) but am not sure.

File: nemoguardrails/guardrails/iorails.py [hunk 13]
```diff
@@ -756,6 +792,24 @@ async def _wrapped_iterator():
...
+                                    if accumulated_tool_calls and not error_emitted:
```
Comment (Pouyanpi) [inline nemoguardrails/guardrails/iorails.py:818]: bare {"tool_calls":...} string is not an openai streaming chunk shape (clients read choices[0].delta.tool_calls); also tool-call args bypass output rails entirely.

**Review by Pouyanpi — APPROVED** [review]
Thanks @tgasser-nv . LGTM! feel free to merge after looking at the following comments.

File: nemoguardrails/guardrails/iorails.py (hunk 4 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/iorails.py]: I changed the design to use a `nonlocal accumulated_tool_calls` which emits the final tool calls without accumulating so this isn't an issue

File: nemoguardrails/guardrails/iorails.py (hunk 5 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/iorails.py]: Same as above, a `nonlocal` rebind is now used instead of mutation

File: nemoguardrails/guardrails/iorails.py (hunk 11 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/iorails.py]: The `ModelEngine.stream_call()`  emits the complete tool-call list exactly once so this contract is maintained. Added a unit-test `test_model_engine_accumulates_tool_calls`  to make sure this contract holds.

File: nemoguardrails/guardrails/iorails.py (hunk 12 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/iorails.py]: Refactored a `_terminal_tool_call_chunk()` helper and described why not what in the comment

File: nemoguardrails/guardrails/iorails.py (hunk 13 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/iorails.py:818]: This is an internal format, the `format_streaming_chunk()` helper takes care of transforming into an OpenAI compatible chunk. 

The tool-call args aren't being validated in this PR, it's in the stacked #2030 . This PR just passes tool calls through in streaming mode

File: nemoguardrails/guardrails/model_engine.py (hunk 8 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/model_engine.py]: Good catch, moved it up to line 234

File: nemoguardrails/guardrails/model_engine.py (hunk 9 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/model_engine.py:271]: Added a TODO so we can refactor a common implementation for IORails and LLMRails for both. The interface is different between these two implementations so it's not a drop-in replacement

File: nemoguardrails/guardrails/model_engine.py (hunk 7 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/model_engine.py]: Removed this comment

File: nemoguardrails/guardrails/model_engine.py (hunk 6 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/model_engine.py]: Deleted

File: nemoguardrails/guardrails/model_engine.py (hunk 10 above)
Comment (tgasser-nv) [reply nemoguardrails/guardrails/model_engine.py:591]: Added a log.warning() if there's an unexpected `finish_reason` while we're still assembling the tool calls. The `test_abnormal_finish_reason_warns_but_still_surfaces_tool_call` test checks for this
